### ISSUE TITLE
Server crashes with `panic: index out of bounds: the len is 2 but the index is 3`

### ISSUE THREAD

**Issue by sushimako** [MaterializeInc/materialize#9504]
### What version of Materialize are you using?

From the shell:

```bash
$ materialized -v
materialized v0.11.1-dev (0d9d13680)
```

### How did you install Materialize?

* [ ] Docker image
* [ ] Linux release tarball
* [x] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [ ] Built from source
* [ ] Materialize Cloud

### What was the issue?
Server crashed with a panic when creating materialized view of query which otherwise can be executed.

### Is the issue reproducible? If so, please provide reproduction instructions.
I wasn't able to recreate a trivial example of the problem, unfortunately. It only occurred on our live-dataset.

The query is equivalent to:
```sql
create materialized view foobar23 as
  select tableB.group_id,
         tableA.id,
         cast(count(*) as int) as value 
    from tableA 
    join tableB on tableB.id = tableA.b_ref 
   where ((tableA.json_col->>'foo' = 'bar7') 
     and (1000 * cast(extract(epoch from tableA.happened_at) as numeric)) <= mz_logical_timestamp() 
     and mz_logical_timestamp() <= (1000 * cast(extract(epoch from tableA.happened_at + interval '90 days') as numeric)))
group by tableB.group_id, tableA.id;
```

Some notes:
 - The `select` query itself can be run and returns the expected results.
 - The statement above works if the last `where` clause is omitted. 

### Please attach any applicable log files.
[9504-materialized.log](https://github.com/MaterializeInc/materialize/files/7684336/9504-materialized.log)

**Comment by asenac** [comment]
Seems related to #9098. Is it possible for `MfpPlan::lower_bounds` to reference intermediate columns added by the `mfp`? If so, then the code below from `MfpPlan::permute` may need to use `permute_map` instead.

```
            for lb in &mut self.lower_bounds {
                lb.permute(permutation);
            }
            for ub in &mut self.upper_bounds {
                ub.permute(permutation);
            }
```

FYI @MaterializeInc/dataflow-optimization

**Comment by uce** [comment]
@sushimako Could you provide the schema of `tableA` and `tableB`? We were not able to reproduce it yet and believe it depends on the schema (including unused columns in the query).

**Comment by umanwizard** [comment]
It is fine to change the names of the columns, in case those are secret; it's just the number and order that matter.

**Comment by asenac** [comment]
I just gave it another try and managed to reproduce the crash with this view:

```
materialize=> create table tablea(b_ref int, some_unused_col int, id int, happened_at timestamp, some_unused_col2 int, json_col json);
materialize=> create table tableb(id int, group_id int, some_unused_col int);
materialize=> create materialized view foobar24 as
  select tableB.group_id,
         tableA.id,
         cast(count(*) as int) as value 
    from tableA 
    join tableB on tableB.id = tableA.b_ref 
   where ((tableA.json_col->>'foo' = 'bar7') 
     and (1000 * cast(extract(epoch from tableA.happened_at) as numeric)) <= mz_logical_timestamp() 
     and mz_logical_timestamp() <= (1000 * cast(extract(epoch from tableA.happened_at) as numeric)))
group by tableB.group_id, tableA.id having tablea.id = tableb.group_id;
CREATE VIEW
materialize=> 
```

Note the missing `+ interval '90 days'` part. That makes `1000 * cast(extract(epoch from tableA.happened_at) as numeric)` a common subexpression in the MFP referenced in the `lower_bound`. Stack trace is:

```
Thread 22 "timely:work-0" hit Breakpoint 1, materialized::handle_panic (panic_info=0x7fffed3bbf68) at src/materialized/src/bin/materialized/main.rs:793
793         let _guard = PANIC_MUTEX.lock();
(gdb) bt
#0  materialized::handle_panic (panic_info=0x7fffed3bbf68) at src/materialized/src/bin/materialized/main.rs:793
#1  0x0000555557fcc5d8 in core::ops::function::Fn::call () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/ops/function.rs:70
#2  0x000055555de19d79 in std::panicking::rust_panic_with_hook () at library/std/src/panicking.rs:628
#3  0x000055555de19800 in std::panicking::begin_panic_handler::{{closure}} () at library/std/src/panicking.rs:521
#4  0x000055555de179d4 in std::sys_common::backtrace::__rust_end_short_backtrace () at library/std/src/sys_common/backtrace.rs:141
#5  0x000055555de19769 in rust_begin_unwind () at library/std/src/panicking.rs:517
#6  0x000055555de3eb61 in core::panicking::panic_fmt () at library/core/src/panicking.rs:101
#7  0x000055555de3eb22 in core::panicking::panic_bounds_check () at library/core/src/panicking.rs:77
#8  0x000055555cfca7b1 in expr::scalar::MirScalarExpr::permute::{{closure}} (e=0x7fffeb18ba40) at src/expr/src/scalar/mod.rs:180
#9  0x000055555cfd4060 in expr::scalar::MirScalarExprVisitor::visit_mut_post::{{closure}} () at src/expr/src/scalar/mod.rs:1067
#10 0x000055555d060bb5 in ore::stack::maybe_grow::{{closure}} () at /home/asenac/work/asenac-materialize-second/src/ore/src/stack.rs:100
#11 0x000055555cfd39b0 in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#12 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize-second/src/ore/src/stack.rs:100
#13 expr::scalar::MirScalarExprVisitor::visit_mut_post (self=0x7fffed3bc510, expr=0x7fffeb18ba40, f=0x7fffed3bc540) at src/expr/src/scalar/mod.rs:1065
#14 0x000055555cfd3dbe in expr::scalar::MirScalarExprVisitor::visit_mut_post::{{closure}}::{{closure}} (e=0x7fffeb18ba40) at src/expr/src/scalar/mod.rs:1066
#15 0x000055555cfd2536 in expr::scalar::MirScalarExprVisitor::visit_mut_children (self=0x7fffed3bc510, expr=0x7fffe37d7800, f=...) at src/expr/src/scalar/mod.rs:1024
#16 0x000055555cfd4044 in expr::scalar::MirScalarExprVisitor::visit_mut_post::{{closure}} () at src/expr/src/scalar/mod.rs:1066
#17 0x000055555d060bb5 in ore::stack::maybe_grow::{{closure}} () at /home/asenac/work/asenac-materialize-second/src/ore/src/stack.rs:100
#18 0x000055555cfd39b0 in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#19 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize-second/src/ore/src/stack.rs:100
#20 expr::scalar::MirScalarExprVisitor::visit_mut_post (self=0x7fffed3bc510, expr=0x7fffe37d7800, f=0x7fffed3bc540) at src/expr/src/scalar/mod.rs:1065
#21 0x000055555cfca486 in expr::scalar::MirScalarExpr::visit_mut_post (warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x55555d039590 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

self=0x7fffe37d7800, warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

f=0x7fffed3bc540) at src/expr/src/scalar/mod.rs:158
warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

#22 0x000055555d0395b5 in expr::scalar::MirScalarExpr::permute (warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

self=0x7fffe37d7800, warning: (Internal error: pc 0x55555d0395b4 in read in psymtab, but not in symtab.)

permutation=...) at src/expr/src/scalar/mod.rs:178
#23 0x000055555d0e24fa in expr::linear::plan::MfpPlan::permute (self=0x7fffed3bcf80, permutation=...) at src/expr/src/linear.rs:1291
#24 0x000055555bc9c91a in dataflow_types::plan::Permutation::permute_mfp_plan (self=0x7fffed3bd000, mfp=0x7fffed3bcf80) at src/dataflow-types/src/plan/mod.rs:990
#25 0x000055555a1c82ac in dataflow::render::context::CollectionBundle<S,repr::row::Row,u64>::as_collection_core::{{closure}} (permutation=...) at src/dataflow/src/render/context.rs:542
#26 0x000055555a1c07a9 in dataflow::render::context::ArrangementFlavor<S,repr::row::Row,T>::flat_map (self=0x7fffe3dc8110, key=..., constructor=...) at src/dataflow/src/render/context.rs:245
#27 0x000055555a1c2511 in dataflow::render::context::CollectionBundle<S,repr::row::Row,T>::flat_map (self=0x7fffed3c04c0, key_val=..., constructor=...) at src/dataflow/src/render/context.rs:377
#28 0x000055555a1c7ced in dataflow::render::context::CollectionBundle<S,repr::row::Row,u64>::as_collection_core (self=0x7fffed3c04c0, mfp=..., key_val=...) at src/dataflow/src/render/context.rs:537
#29 0x000055555a1de4d0 in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan (self=0x7fffed3ecd10, plan=..., scope=0x7fffed3ef6f0, worker_index=0) at src/dataflow/src/render/mod.rs:460
#30 0x000055555a1dcc2a in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan (self=0x7fffed3ecd10, plan=..., scope=0x7fffed3ef6f0, worker_index=0) at src/dataflow/src/render/mod.rs:552
#31 0x0000555558581bf6 in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan::{{closure}} (input=...) at src/dataflow/src/render/mod.rs:500
#32 0x0000555558ec8ffa in core::iter::adapters::map::map_fold::{{closure}} (acc=(), elt=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/iter/adapters/map.rs:84
#33 0x00005555590bd9e3 in core::iter::traits::iterator::Iterator::fold (self=..., init=(), f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/iter/traits/iterator.rs:2170
#34 0x0000555558e5bc88 in <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold (self=..., init=(), g=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/iter/adapters/map.rs:124
#35 0x0000555558e940cf in core::iter::traits::iterator::Iterator::for_each (self=..., f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/iter/traits/iterator.rs:733
#36 0x000055555a608745 in <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend (self=0x7fffed3d6350, iterator=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/vec/spec_extend.rs:40
#37 0x000055555a488225 in <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter (iterator=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/vec/spec_from_iter_nested.rs:56
#38 0x000055555a583158 in alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter (iterator=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/vec/source_iter_marker.rs:41
#39 0x000055555a5eb340 in <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter (iter=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/vec/mod.rs:2486
#40 0x0000555558e90d4c in core::iter::traits::iterator::Iterator::collect (self=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/iter/traits/iterator.rs:1745
#41 0x000055555a1dd983 in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan (self=0x7fffed3ecd10, plan=..., scope=0x7fffed3ef6f0, worker_index=0) at src/dataflow/src/render/mod.rs:498
#42 0x000055555a1dc97c in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan (self=0x7fffed3ecd10, plan=..., scope=0x7fffed3ef6f0, worker_index=0) at src/dataflow/src/render/mod.rs:517
#43 0x000055555a1dc6d9 in dataflow::render::<impl dataflow::render::context::Context<G,repr::row::Row,u64>>::render_plan (self=0x7fffed3ecd10, plan=..., scope=0x7fffed3ef6f0, worker_index=0) at src/dataflow/src/render/mod.rs:480
#44 0x000055555a1db43b in dataflow::render::<impl dataflow::render::context::Context<timely::dataflow::scopes::child::Child<G,<G as timely::dataflow::scopes::ScopeParent>::Timestamp>,repr::row::Row,u64>>::build_object (self=0x7fffed3ecd10, scope=0x7fffed3ef6f0, object=...) at src/dataflow/src/render/mod.rs:343
#45 0x0000555558580d4b in dataflow::render::build_dataflow::{{closure}}::{{closure}} (region=0x7fffed3ef6f0) at src/dataflow/src/render/mod.rs:262
#46 0x0000555559136bcf in <timely::dataflow::scopes::child::Child<G,T> as timely::dataflow::scopes::Scope>::scoped (self=0x7fffed3f02b0, name=..., func=...) at /home/asenac/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/fa63c3a/timely/src/dataflow/scopes/child.rs:129
#47 0x00005555591380b2 in timely::dataflow::scopes::Scope::region_named (self=0x7fffed3f02b0, name=..., func=...) at /home/asenac/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/fa63c3a/timely/src/dataflow/scopes/mod.rs:187
#48 0x00005555585818b4 in dataflow::render::build_dataflow::{{closure}} (scope=0x7fffed3f08f0) at src/dataflow/src/render/mod.rs:205
#49 0x000055555858688e in timely::worker::Worker<A>::dataflow_core (self=0x7fffed3f8c40, name=..., logging=..., resources=0x1, func=...) at /home/asenac/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/fa63c3a/timely/src/worker.rs:639
#50 0x000055555857fe03 in dataflow::render::build_dataflow (timely_worker=0x7fffed3f8c40, render_state=0x7fffed3f8290, dataflow=..., now=..., source_metrics=0x7fffed3f85a0, sink_metrics=0x7fffed3f8608) at src/dataflow/src/render/mod.rs:200
#51 0x0000555558ca1417 in dataflow::server::Worker<A>::handle_command (self=0x7fffed3f8288, cmd=...) at src/dataflow/src/server.rs:678
#52 0x0000555558c9a7da in dataflow::server::Worker<A>::run (self=0x7fffed3f8288) at src/dataflow/src/server.rs:487
#53 0x0000555558c962e8 in dataflow::server::serve::{{closure}} (timely_worker=0x7fffed3f8c40) at src/dataflow/src/server.rs:131
#54 0x0000555558ca7fc9 in timely::execute::execute::{{closure}} (allocator=...) at /home/asenac/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/fa63c3a/timely/src/execute.rs:285
#55 0x0000555559f2c1cb in timely_communication::initialize::initialize_from::{{closure}} () at /home/asenac/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/fa63c3a/communication/src/initialize.rs:316
#56 0x00005555585686f0 in std::sys_common::backtrace::__rust_begin_short_backtrace (f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/sys_common/backtrace.rs:125
#57 0x0000555558aa8384 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/thread/mod.rs:481
#58 0x0000555559f208c4 in <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=()) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/panic/unwind_safe.rs:271
#59 0x0000555559cd4602 in std::panicking::try::do_call (data=0x7fffed3f9828 "\001\000") at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panicking.rs:403
#60 0x0000555559cdc0fb in __rust_try ()
#61 0x0000555559cd34e7 in std::panicking::try (f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panicking.rs:367
#62 0x00005555584fa554 in std::panic::catch_unwind (f=...) at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/panic.rs:129
#63 0x0000555558aa7693 in std::thread::Builder::spawn_unchecked::{{closure}} () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/std/src/thread/mod.rs:480
#64 0x000055555887f55f in core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/ops/function.rs:227
#65 0x000055555de20b93 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/boxed.rs:1636
#66 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/alloc/src/boxed.rs:1636
#67 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:106
#68 0x00007ffff7f94609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#69 0x00007ffff7d2f293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

**Comment by asenac** [comment]
Simpler repro case:

```
materialize=> create materialized view foobar24 as                                                                                    
  select tableA.id 
    from tableA 
                
   where ((1000 * cast(extract(epoch from tableA.happened_at) as numeric)) <= mz_logical_timestamp() 
     and mz_logical_timestamp() <= (1000 * cast(extract(epoch from tableA.happened_at) as numeric)));
CREATE VIEW
materialize=>
```

**Comment by asenac** [comment]
As discussed in slack, some permutation needs to be applied to the upper and lower bounds, because they may reference both a CSE and an input column. If we don't apply any permutation we would just move the crash to the execution time.

The following example contains a view where all input columns are shifted 1 position and where the upper bound references both a CSE column and an input column:

```
materialize=> create table tablea(happened_at timestamp, json_col json, offs numeric, id int);
CREATE TABLE
materialize=> create index a on tablea(id);
CREATE INDEX
materialize=> create materialized view foobar26 as                                                                                    
  select tableA.id 
    from tableA                 
   where ((1000 * cast(extract(epoch from tableA.happened_at) as numeric)) <= mz_logical_timestamp() 
     and mz_logical_timestamp() <= (1000 * cast(extract(epoch from tableA.happened_at) as numeric) + offs )) and id = 10;
CREATE VIEW
materialize=>
```

If we just avoid doing any permutation the following crash happens when we add the first row to the table:

```
(gdb) bt 10
#0  materialized::handle_panic (panic_info=0x7fffed3efec8) at src/materialized/src/bin/materialized/main.rs:793
#1  0x0000555557fcd5d8 in core::ops::function::Fn::call () at /rustc/09c42c45858d5f3aedfa670698275303a3d19afa/library/core/src/ops/function.rs:70
#2  0x000055555de1ad79 in std::panicking::rust_panic_with_hook () at library/std/src/panicking.rs:628
#3  0x000055555de1a800 in std::panicking::begin_panic_handler::{{closure}} () at library/std/src/panicking.rs:521
#4  0x000055555de189d4 in std::sys_common::backtrace::__rust_end_short_backtrace () at library/std/src/sys_common/backtrace.rs:141
#5  0x000055555de1a769 in rust_begin_unwind () at library/std/src/panicking.rs:517
#6  0x000055555de3fb61 in core::panicking::panic_fmt () at library/core/src/panicking.rs:101
#7  0x000055555de3fb22 in core::panicking::panic_bounds_check () at library/core/src/panicking.rs:77
#9  0x000055555d0af486 in <T as expr::scalar::func::LazyUnaryFunc>::eval (self=0x7fffe8c0d309, datums=..., temp_storage=0x7fffed3f3600, a=0x7fffeb79f000) at src/expr/src/scalar/func.rs:3090
(More stack frames follow...)

(gdb) 
```

**Comment by alex-hunt-materialize** [comment]
We have a cloud deployment running v0.10.0 which seems to be hitting this too.

**Comment by antiguru** [comment]
We believe this issue was fixed in #9524, which will be included in the upcoming release 0.12. Thanks again for reporting the issue and everyone who help to reproduce the problem!
